zerver/tests/test_auth_backends.py

```diff
@@ -850,3 +877,368 @@ def test_login_success_under_subdomains(self):
+            self.assertIs(user_profile, None)
+
+    def test_login_success_under_subdomains(self):
+        # type: () -> None
+        self.mock_ldap.directory = {
+            'uid=hamlet,ou=users,dc=zulip,dc=com': {
+                'userPassword': 'testing'
+            }
+        }
+        with self.settings(
+                REALMS_HAVE_SUBDOMAINS=True,
+                LDAP_APPEND_DOMAIN='zulip.com',
+                AUTH_LDAP_BIND_PASSWORD='',
+                AUTH_LDAP_USER_DN_TEMPLATE='uid=%(user)s,ou=users,dc=zulip,dc=com'):
+            user_profile = self.backend.authenticate('hamlet@zulip.com', 'testing')
+            self.assertEqual(user_profile.email, 'hamlet@zulip.com')
+
+class TestZulipLDAPUserPopulator(ZulipTestCase):
+    def test_authenticate(self):
+        # type: () -> None
+        backend = ZulipLDAPUserPopulator()
+        result = backend.authenticate('hamlet@zulip.com', 'testing')  # type: ignore # complains that the function does not return any value!
+        self.assertIs(result, None)
+
+class TestZulipAuthMixin(ZulipTestCase):
+    def test_get_user(self):
+        # type: () -> None
+        backend = ZulipAuthMixin()
+        result = backend.get_user(11111)
+        self.assertIs(result, None)
+
+class TestPasswordAuthEnabled(ZulipTestCase):
+    def test_password_auth_enabled_on_production(self):
+        # type: () -> None
+        with self.settings(PRODUCTION=True):
+            realm = Realm.objects.get(domain='zulip.com')
+            self.assertFalse(password_auth_enabled(realm))
+
+    def test_password_auth_enabled_for_ldap(self):
+        # type: () -> None
+        with self.settings(AUTHENTICATION_BACKENDS=('zproject.backends.ZulipLDAPAuthBackend',)):
+            realm = Realm.objects.get(domain='zulip.com')
+            self.assertTrue(password_auth_enabled(realm))
+
+class TestMaybeSendToRegistration(ZulipTestCase):
+    def test_sso_only_nonexisting_user(self):
+        # type: () -> None
+        rf = RequestFactory()
+        request = rf.get('/')
+        request.session = {}
+        request.user = None
+
+        class Form(object):
+            def is_valid(self):
+                # type: () -> bool
+                return True
+
+        with self.settings(ONLY_SSO=True):
+            with mock.patch('zerver.views.auth.create_homepage_form', return_value=Form()):
+                self.assertEqual(PreregistrationUser.objects.all().count(), 0)
+                result = maybe_send_to_registration(request, 'hamlet@zulip.com')
+                self.assertEqual(result.status_code, 302)
+                self.assertEqual(PreregistrationUser.objects.all().count(), 1)
+
+    def test_sso_only_existing_user(self):
+        # type: () -> None
+        rf = RequestFactory()
+        request = rf.get('/')
+        request.session = {}
+        request.user = None
+
+        class Form(object):
+            def is_valid(self):
+                # type: () -> bool
+                return True
+
```

Comment:
Why do we need all this setup code for this test and the previous one?  Maybe worth a comment complaining?